### PR TITLE
Add new blog on null safety using spring boot

### PR DESCRIPTION
**Pull request by s4ik4t**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [x] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [x] the article explains *why* and not only *how*
- [x] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [x] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [x] the text is conversational ("we" instead of "you", easy language)
- [x] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [x] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [x] the text uses active voice instead of passive voice
- [x] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [x] file names and variables are highlighted as `code`

## Structure
- [x] the article has a short intro before the first headline
- [x] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [x] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [x] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [x] code examples are introduced with a ":" and explained below the code example
- [x] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+---
+authors: [tom]
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:2]: Please add yourself in the file data/authors.yml and then put your name here so that the article is attributed to you

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+title: "Null-safety annotations using Spring Boot"
+categories: ["Spring Boot"]
+date: 2022-01-16T00:00:00
+excerpt: "A guide on how to write null-safe code using Spring annotations"
+url: spring-boot-null-safety-annotations
+---
+
+## Introduction
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: We don't need a header for the introduction, I removed it.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+Since `NullPointerException` is a runtime exception, it would be hard to figure out such cases during code compilation. Java's type system does not have a way to quickly eliminate the dangerous null object references. 
+
+But, what if I tell you there are helpful approaches to write null-safe code. In particular, if you are using the Spring Framework. Wouldn't that be great news?
+
+Luckily, Spring Framework offers some annotations to solve this exact problem. 
+In this article, we will learn how to use these annotations to write null-safe code using [Spring Boot](https://reflectoring.io/categories/spring-boot/).
+
+## Null-Safety Annotations in Spring
+
+Under the `org.springframework.lang` Spring core package, there are 4 such annotations, `@NonNull`, `@NonNullFields`, `@Nullable`, and `@NonNullApi`. These are [meta-annotations](https://dzone.com/articles/what-are-meta-annotations-in-java) that use the famous (even though dormant at present) [JSR-305](https://jcp.org/en/jsr/detail?id=305) annotations underneath.
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: I removed a few links across the article, because the text was too "busy" with links.

In this case, I also removed the mention of the JSR, because it only distracts the reader. If you want to mention it, you can re-add it, but please explain in a sentence what it does (maybe use an [aside](https://reflectoring.io/contribute/writing-guide/#asides) to explain it).

Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Why are these annotations meta annotations? We're using them directly on our code later, which means they're not just meta annotations. I removed the sentence, but feel free to re-add it, but with an explanation.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  //...
+}
+```
+Now, if we accidentally try to set the value of `id` as null anywhere in the code, the IDE will show a compilation warning:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Pretty awesome, isn't it?
+
+**The `NonNull` annotation can be used at the method, parameter, or field level.**
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Could you maybe add a [table](https://reflectoring.io/contribute/writing-guide/#tables) to each section that explains what happens when you put the respective annotation on a method, parameter, field, package?

A table is also a nice way to break up the text a bit :)

Something like:

| Annotated element | Effect |
|--------------------------|---------|
| field | Shows a warning when the field is null |
| parameter | ... |

Ideally also show a small code example for each of these cases, just to make it extra clear to the reader how to use it.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+```java
+@NonNullApi
+package io.reflectoring.nullsafety;
+
+import org.springframework.lang.NonNullApi;
+```
+
+We can now see the IDE is warning us about the non-nullable parameters:
+
+{{% image alt="IDE warning for NonNullApi" src="images/posts/spring-boot-null-safety-annotations/nonnullapi-ide-warning.png" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:112]: Please add a code example that shows how you use the `@NonNullApi` annotation. From just the warning I can guess how it works, but I don't know for sure :)

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+**Popular IDEs like [Eclipse](https://www.eclipse.org/ide/) and [IntelliJ IDEA](https://www.jetbrains.com/idea/) can understand these annotations.** It can warn the developers of potential issues during compile time. 
+
+We are going to use IntelliJ IDEA in this tutorial. Let us find out more with some code examples.
+
+To create the base project, we have taken help from the [Spring Initializr](https://start.spring.io/). The Spring Boot starter is all we need, no need to add any extra dependencies.
+
+## `@NonNull`
+
+Let's use a plain `Employee` class to explain the annotations:
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:37]: Please pull the explanation of the Employee class into its own section titled "example code" or similar, because you're using it in all the sections below and a reader might want to skip directly to the `@Nullable`  and then doesn't know what the `pastEmployment` field is. If it's in a separate action, the reader can just look at the table of contents to navigate to the "example code" section.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+**The `NonNull` annotation can be used at the method, parameter, or field level.** 
+
+At this point, you might be thinking, what if a class has more than one non-null field. Would it not be too wordy if we have to add a `@NonNull` annotation before each of these? 
+
+We can solve this problem by using the [`@NonNullFields`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullFields.html) annotation. Let us see how in the next section.
+
+## `@NonNullFields`
+
+Let us create a `package-info.java` file to apply the non-null field checks at the package level. This file will contain the root package name with `@NonNullFields` annotation:
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:80]: This makes it seem that `@NonNullFields` is only applicable to packages. That's not true, is it? Please add a table like mentioned in the comment above.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  \\...
+```
+
+## Conclusion
+
+These annotations are indeed a boon for the Java programmers to reduce the possibility of a `NullPointerException` arising during runtime. But, please bear in mind this does not guarantee complete null safety.
+
+**However, please note that not all development tools can show these warnings. If you don't see the compilation warning, check the compiler settings in IntelliJ.**
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: This is important information that should not be hidden in the conclusion! Can you extract this into its own section at the start of the article? And maybe also find out and explain how to enable this in Eclipse?

Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Also, is there a way to evaluate these annotations during a Maven or Gradle build? That would be interesting...and certainly worth another section in this article :)

**Review by thombergs — CHANGES_REQUESTED** [review]
Good first article! I left a few comments to make this article more consistent and informative.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 3 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Sounds good to me, thanks!

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 7 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:80]: `@NonNullFields` is meant for packages only. 
[Source](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullFields.html)

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 6 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:37]: I should have thought about that, thanks!

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 4 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Great suggestion! I have added a table at the end of each section to summarize the annotation behavior. Is it OK if we skip adding a code snippet in the table to make it concise (because for a couple of annotations it would require an extra clarification)?

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 5 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:137]: Fair enough. I have now updated this section.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 8 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: Done, thanks!

Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md]: I found the [SpotBugs](https://spotbugs.readthedocs.io/en/latest/introduction.html) Maven/Gradle Plugin is able to offer some help. Added a section on that.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+```groovy
+dependencies {
+  spotbugsPlugins 'com.h3xstream.findsecbugs:findsecbugs-plugin:1.11.0'
+}
+
+spotbugs {
+  toolVersion = '4.5.3'
+}
+```
+Once the project is updated, we can run the check using the `gradle check` command.
```
Comment (thombergs) [inline content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:252]: Last question before publishing. 

When I read this section about SpotBugs, I'm asking myself how this is connected to the Null annotations. Does SpotBugs evaluate these annotations? Or does SpotBugs have some other nullability checks? Would be good to explain that in the text. If SpotBugs doesn't evaluate the annotations, we should remove the section again, because this article is about the annotations.

**Review by thombergs — CHANGES_REQUESTED** [review]
Nice! Looks good! One last question in a comment.

File: content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md (hunk 9 above)
Comment (s4ik4t) [reply content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md:259]: Hi Tom, thanks, I have slightly modified that section. Please have a look. SpotBugs could be another potential topic for us :)

**Review by thombergs — APPROVED** [review]
